adserver/views.py

```diff
@@ -766,6 +772,161 @@ def get_success_url(self):
+                Image = namedtuple("Image", ["url"])
+
+                for ad in ads:
+                    ad_objects.append(
+                        Advertisement(
+                            name=ad["name"],
+                            image=Image(ad["image_url"]),
+                            live=ad["live"],
+                            link=ad["link"],
+                            headline=ad["headline"],
+                            content=ad["content"],
+                            cta=ad["cta"],
+                        )
+                    )
+
+                context = self.get_context_data()
+                context["preview_ad_type"] = preview_ad_type
+                context["preview_ads"] = ad_objects
+                context["signed_advertisements"] = signer.sign_object(ads)
+                return self.render_to_response(context)
+            else:
+                return self.form_invalid(form)
+        elif "signed_advertisements" in self.request.POST:
+            # Process previewed ads and save them
+            signer = self.get_signer()
+
+            try:
+                ads = signer.unsign_object(
+                    self.request.POST["signed_advertisements"],
+                    max_age=self.MAXIMUM_SIGNED_TIME_LENGTH,
+                )
+            except signing.BadSignature:
+                messages.error(
+                    self.request,
+                    _("Upload expired or invalid. Please upload ads again."),
+                )
+                return redirect(self.get_error_url())
+
+            # Save the ads which have already been validated
+            for ad in ads:
+                ad_object = Advertisement(
+                    flight=self.flight,
+                    name=ad["name"],
+                    slug=self.get_ad_slug(ad["name"]),
+                    image=ImageFile(
+                        default_storage.open(ad["image_path"]), name=ad["image_name"]
+                    ),
+                    live=ad["live"],
+                    link=ad["link"],
+                    headline=ad["headline"],
+                    content=ad["content"],
+                    cta=ad["cta"],
+                )
+                ad_object.save()
+                ad_object.ad_types.set(
+                    self.flight.campaign.allowed_ad_types(exclude_deprecated=True)
+                )
+
+            messages.success(
+                self.request,
+                _("Successfully uploaded %(num_ads)s ads") % {"num_ads": len(ads)},
+            )
+            return redirect(self.get_success_url())
+        else:
+            return redirect(self.get_error_url())
+
+    def get_ad_slug(self, name):
+        slug = slugify(f"{name}")
+
+        while Advertisement.objects.filter(slug=slug).exists():
+            random_chars = get_random_string(8)
+            slug = slugify(f"{name}-{random_chars}")
+
```

Comment:
I feel like this logic also already exists on the normal ad upload? IIRC we attached a date to it there -- should probably break that out into a util as well?